File: apps/scala/design-pattern/src/IteratorPattern.scala
// Iterator パターン
// Iterable trait と given/using による Ordering を活用する

package designpattern.iterator

case class Account(name: String, balance: Double)

object Account:
  // given で暗黙の Ordering を定義
  given Ordering[Account] = Ordering.by(_.balance)

class Portfolio(private val accounts: List[Account]) extends Iterable[Account]:
  override def iterator: Iterator[Account] = accounts.iterator

  def sortedByBalance: List[Account] =
    import Account.given
    accounts.sorted

  def sortedByName: List[Account] =
    accounts.sortBy(_.name)

  def totalBalance: Double =
    accounts.map(_.balance).sum

  def findByName(name: String): Option[Account] =
    accounts.find(_.name == name)

  def filterBy(predicate: Account => Boolean): List[Account] =
    accounts.filter(predicate)

File: apps/haskell/design-pattern/src/IteratorPattern.hs
-- | Iterator パターン
-- Haskell ではリストが標準のイテレータ。
-- Data.Map や独自の走査関数で多様な反復処理を示す。
module IteratorPattern
  ( Portfolio(..)
  , Account(..)
  , AccountType(..)
  , newPortfolio
  , addAccount
  , totalBalance
  , filterByType
  , accountNames
  , sortByBalance
  , mergePortfolios
  ) where

import Data.List (sortBy)
import Data.Ord (comparing)

-- | 口座種別
data AccountType = Savings | Checking | Investment
  deriving (Show, Eq, Ord)

-- | 口座
data Account = Account
  { acctName    :: String
  , acctType    :: AccountType
  , acctBalance :: Double
  } deriving (Show, Eq)

-- | ポートフォリオ: 口座のコレクション
newtype Portfolio = Portfolio { accounts :: [Account] }
  deriving (Show, Eq)

-- | 空のポートフォリオ
newPortfolio :: Portfolio
newPortfolio = Portfolio []

-- | 口座を追加
addAccount :: Account -> Portfolio -> Portfolio
addAccount a (Portfolio as) = Portfolio (as ++ [a])

-- | 残高合計（fold）
totalBalance :: Portfolio -> Double
totalBalance = sum . map acctBalance . accounts

-- | 種別でフィルタ（filter）
filterByType :: AccountType -> Portfolio -> [Account]
filterByType t = filter (\a -> acctType a == t) . accounts

-- | 口座名一覧（map）
accountNames :: Portfolio -> [String]
accountNames = map acctName . accounts

-- | 残高順にソート
sortByBalance :: Portfolio -> [Account]
sortByBalance = sortBy (comparing acctBalance) . accounts

-- | 2 つのポートフォリオを統合
mergePortfolios :: Portfolio -> Portfolio -> Portfolio
mergePortfolios (Portfolio a) (Portfolio b) = Portfolio (a ++ b)
